Lib/typing.py

```diff
@@ -2415,15 +2415,17 @@ def utf8(value: str) -> bytes: ...
       def utf8(value):
           # implementation goes here
 
-    Overload definition can be retrieved at runtime using the
+    The overloads for a function can be retrieved at runtime using the
     get_overloads() function.
     """
     try:
-        existing = get_overloads(func)
-    except AttributeError:
+        key = _get_key_for_callable(func)
```

Comment:
What if you inlined this call?

Reply:
That helps a little, though not as much as @AlexWaygood's suggestion to use a tuple as the key.

I had another idea: we perform the _get_firstlineno() call below to avoid leaking unbounded memory if someone keeps reloading a module with overloads. But we don't really need to check for that every single time, we can get away with just doing it say every time the number of overloads reaches a multiple of 16. That cuts down the overhead to about 8x.

Reply:
That idea is unsafe, because we detect reloads by checking if the lineno of the current overload <= the lineno of the last one, and if we only check that every N times, we may miss the window where we return back to the beginning. However, I think it's sound to skip the lineno check if the total number of overloads <= N, and set N to say 16. Most real-world functions have fewer than 16 overloads (I hope!), and if a module is repeatedly reloaded with new overloads being registered, they'll still eventually be cleared up.

Reply:
Okay, so if a function has 3 overloads and the module is (re)loaded three times, for a total of 9 `@overload` calls (i.e., less than 9), will `get_overloads()` return 3 or 9 function objects?

Reply:
It will return 9. I think that's fine; we don't need to guarantee any specific behavior in the face of things like reloads, just avoid leaking unbounded amounts of memory.

Reply:
It's not only reloads, more a realistic case would be nested functions being defined multiple times.

Reply:
Ew. That would be hard to explain to the users of the API (who probably aren't so in tune with our worry about memory leaks) and we'd likely get plenty of bug reports of the form "my function has three overloads, why do I see six?"

If you can clean this up when the overload count goes over 16, why can't you clean it up in `get_overloads()`?

(Also, why 16? In this day and age, numbers that are based on multiples of two are rarely special, and users are more likely to remember/understand numbers that are small multiples of powers of 10. :-)

Reply:
Good point, I'll add cleanup to get_overloads().

I guess I chose 16 because I assumed % 16 is cheaper to compute. Not sure that's even true in Python code. But we're not even using % any more, so I'll change it to 20.

Nested functions wouldn't actually work in all cases. If you dynamically generate functions with overloads and call get_overloads(), you'll get the overloads for the most recently defined function, which isn't necessarily the one you want. To get around it you'd have to mess with `__name__`.

Reply:
Timing with this variant (about to push):

```
In [6]: %timeit old_overload(f)
45.7 ns ± 0.43 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %timeit overload(f)
286 ns ± 1.47 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [9]: 286 / 46
Out[9]: 6.217391304347826
```

Reply:
> Nested functions wouldn't actually work in all cases. If you dynamically generate functions with overloads and call get_overloads(), you'll get the overloads for the most recently defined function, which isn't necessarily the one you want. To get around it you'd have to mess with `__name__`.

I suppose that's okay, but it does remind me that the overload dict may keep overloaded functions alive beyond the implementation's lifetime. And that in turn means that we have effectively immortalized the globals of all overloaded functions. Is *that* okay? (I had some trouble demonstrating this effect until I remembered that *every* (global) function keeps its globals alive; `gc.collect()` shows it though.)